### PR TITLE
Automate include grouping order in .clang-format

### LINKED ISSUE #15103: Automate C++ include file grouping and ordering using clang-format
Currently the C++ include files in the RAPIDS codebase are not consistently ordered and grouped.
Inconsistent ordering makes it harder to maintain the codebase, to understand the dependencies,
and to automate changes.

Using clang-format to automate the ordering and grouping of include files will not only add
consistent style, but it will ease automation of changes. For example, We are undertaking a
refactoring of RMM that will replace `rmm::mr::device_memory_resource*` with
`rmm::device_async_resource-ref` everywhere in RAPIDS (not just cuDF). This requires adding an
include to MANY files across multiple RAPIDS repos. Getting the location of this include correct
everywhere is very difficult without automatic grouping of headers.

I propose to use clang-format's `IncludeCategories` settings to automate the ordering and
grouping of include files in the C++ codebase consistently.

Proof-of-concept PRs have been created for RMM and cuDF:
 - cuDF: https://github.com/rapidsai/cudf/pull/15063
 - RMM: https://github.com/rapidsai/rmm/pull/1463

The ordering used for cuDF is given by the following settings.

```yaml
IncludeBlocks: Regroup
IncludeCategories:
  - Regex:           '^"' # quoted includes
    Priority:        1
  - Regex:           '^<(benchmarks|tests)/' # benchmark includes
    Priority:        2
  - Regex:           '^<cudf_test/' # cuDF includes
    Priority:        3
  - Regex:           '^<cudf/' # cuDF includes
    Priority:        4
  - Regex:           '^<(nvtext|cudf_kafka)' # other libcudf includes
    Priority:        5
  - Regex:           '^<(cugraph|cuml|cuspatial|raft|kvikio)' # Other RAPIDS includes
    Priority:        6
  - Regex:           '^<rmm/' # RMM includes
    Priority:        7
  - Regex:           '^<(thrust|cub|cuda)/' # CCCL includes
    Priority:        8
  - Regex:           '^<(cooperative_groups|cuco|cuda.h|cuda_runtime|device_types|math_constants|nvtx3)' # CUDA includes
    Priority:        8
  - Regex:           '^<.*\..*' # other system includes (e.g. with a '.')
    Priority:        9
  - Regex:           '^<[^.]+' # STL includes (no '.')
    Priority:        10
```

### PR DESCRIPTION
**Pull request by harrism**
Replaces #14993 
Closes #15103

## Description
This uses the `IncludeCategories` settings in .clang-format to attempt to enforce our documented `#include` order in libcudf. See https://docs.rapids.ai/api/libcudf/stable/developer_guide

I realize that there was a [previous attempt at this](https://github.com/rapidsai/cudf/pull/12760) by @bdice that met with some resistance. Reading it, I wouldn't say it was vetoed; rather, reviewers requested something much simpler. I have a few reasons to attempt this again.

1. To make a separate task much easier. We are undertaking a refactoring of RMM that will replace `rmm::mr::device_memory_resource*` with `rmm::device_async_resource-ref` everywhere in RAPIDS (not just cuDF). This requires adding an include to MANY files. Getting the location of the include right everywhere is very difficult without automatic grouping of headers. I started out writing a bash script to do this before realizing clang-format has the necessary feature. And I realized that my script would never properly handle [files like this](https://github.com/rapidsai/raft/blob/branch-24.04/cpp/bench/ann/src/raft/raft_cagra_wrapper.h).
2. To increase velocity. Everywhere in RAPIDS that we have automated code standard/style/formatting/other, the benefits to velocity have outweighed the costs. To paraphrase @bdice, $auto \nearrow \rightarrow \mu \searrow \rightarrow v \nearrow$
3. The previous PR #12760 had nearly 50 categories of headers. There was no way this could be applied universally across RAPIDS repos. My proposal has 10 categories. I tried to reduce it further but realized that it wouldn't be much less configuration to maintain, so I stopped at 10.

Note that one of the ways that having few categories can work while still maintaining clear groups is that this PR updates many files to use quotes ("") instead of angle brackets (<>) for local cuDF headers that do not live in `cudf/cpp/include`. With our "near to far" include ordering policy, these are arguably the nearest files, and using quotes allows us to have our first category simply check for quotes. These files will be grouped and sorted without blank lines, but in practice this does not lose clarity because typically headers from more than two directories are not included from the same file. The downside of this change is I don't yet know how to automatically enforce it. I hope that when developers accidentally use <> for internal includes that don't start with (e.g.) "cudf", they will be grouped one of the lowest priority categories, and perhaps this will induce them to switch to "" to get the headers listed at the top. The rule is simple: if it's in libcudf but not in `cpp/include/cudf`, then use quotes. For **everything** else, use angle brackets.

Other than headers from RAPIDS repos, we have a group for all CCCL/CUDA headers, a group for all other headers that have a file extension, and a final group for all files that have no file extension (e.g. STL).

Below I'm listing the (fairly simple, in my opinion) .clang-format settings for this PR. Note that categories 2-5 will require tweaking for different RAPIDS repos. 

Some may ask why I ordered `cudf_test` headers before `cudf` headers. I tried both orders, and putting `cudf_test` first generated significantly fewer changes in the PR, meaning that it's already the more common ordering (I suppose `cudf_test` is closer to the files that include it, since they are libcudf tests).

I've opened a similar PR for RMM with only 5 groups. https://github.com/rapidsai/rmm/pull/1463

CC @davidwendt @vyasr @wence- @GregoryKimball for feedback

@isVoid contributed to this PR via pair programming.

```
IncludeBlocks: Regroup
IncludeCategories:
  - Regex:           '^"' # quoted includes
    Priority:        1
  - Regex:           '^<(benchmarks|tests)/' # benchmark includes
    Priority:        2
  - Regex:           '^<cudf_test/' # cuDF includes
    Priority:        3
  - Regex:           '^<cudf/' # cuDF includes
    Priority:        4
  - Regex:           '^<(nvtext|cudf_kafka)' # other libcudf includes
    Priority:        5
  - Regex:           '^<(cugraph|cuml|cuspatial|raft|kvikio)' # Other RAPIDS includes
    Priority:        6
  - Regex:           '^<rmm/' # RMM includes
    Priority:        7
  - Regex:           '^<(thrust|cub|cuda)/' # CCCL includes
    Priority:        8
  - Regex:           '^<(cooperative_groups|cuco|cuda.h|cuda_runtime|device_types|math_constants|nvtx3)' # CUDA includes
    Priority:        8
  - Regex:           '^<.*\..*' # other system includes (e.g. with a '.')
    Priority:        9
  - Regex:           '^<[^.]+' # STL includes (no '.')
    Priority:        10
```

## Checklist
- [X] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [X] New or existing tests cover these changes.
- [X] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by wence-** [comment]
As mentioned on the other PR, I think this is great, thanks for doing it! I just had one bikeshed colour policy question:

```
  - Regex:           '^<cudf/' # cuDF includes
    Priority:        4
```

Is it desirable to have a separation of detail API headers from public ones? (so `'^<cudf/detail'` with a separate priority from `'^<cudf/'`) This would work because matching is done top down, so we would just put the detail regex before the public API regex in the list.

**Comment by davidwendt** [comment]
> As mentioned on the other PR, I think this is great, thanks for doing it! I just had one bikeshed colour policy question:
> 
> ```
>   - Regex:           '^<cudf/' # cuDF includes
>     Priority:        4
> ```
> 
> Is it desirable to have a separation of detail API headers from public ones? (so `'^<cudf/detail'` with a separate priority from `'^<cudf/'`) This would work because matching is done top down, so we would just put the detail regex before the public API regex in the list.

I don't think we need this extra group.

**Comment by harrism** [comment]
Moving out of draft

File: cpp/src/io/parquet/reader_impl.cpp [hunk 1]
```diff
@@ -15,6 +15,7 @@
  */
 
 #include "reader_impl.hpp"
+
```
Comment (davidwendt) [inline cpp/src/io/parquet/reader_impl.cpp:18]: I wonder what happened here.

File: cpp/src/join/mixed_join_size_kernels_semi.cu [hunk 2]
```diff
@@ -26,6 +22,10 @@
 
 #include <cub/cub.cuh>
 
+#include <join/join_common_utils.cuh>
+#include <join/join_common_utils.hpp>
+#include <join/mixed_join_common_utils.cuh>
```
Comment (davidwendt) [inline cpp/src/join/mixed_join_size_kernels_semi.cu:27]: This move seems wrong. Perhaps they are meant to be quoted?

File: cpp/src/utilities/logger.cpp [hunk 3]
```diff
@@ -14,10 +14,11 @@
  * limitations under the License.
  */
 
+#include "spdlog/sinks/stdout_sinks.h"
```
Comment (davidwendt) [inline cpp/src/utilities/logger.cpp:17]: This should match with its companion at line 22.
https://github.com/rapidsai/cudf/pull/15063/files#diff-906fc8cdd3162b1f4778c3e85d9e72505ff171ccaaaab5c4a8f704f2cef7637eR22

**Review by wence- — COMMENTED** [review]
Up to the start of the test files.

File: cpp/include/cudf/fixed_point/temporary.hpp [hunk 4]
```diff
@@ -15,9 +15,6 @@
  */
 
 #pragma once
-// To avoid https://github.com/NVIDIA/libcudacxx/issues/460
-// in libcudacxx with CTK 12.0/12.1
-#include <cuda_runtime.h>
```
[the remark this answers is not in the document]
Comment (bdice) [reply cpp/include/cudf/fixed_point/temporary.hpp:20]: That is correct but probably better for a separate PR.

[the remark this answers is not in the document]
Comment (bdice) [reply cpp/include/cudf/fixed_point/temporary.hpp:20]: I noted this on the previous PR that was closed, I think. Can we double check that my previous comments were addressed (or deferred for a later issue/PR)?

[the remark this answers is not in the document]
Comment (bdice) [reply cpp/include/cudf/fixed_point/temporary.hpp:20]: Nevermind. It was addressed here. I misread the diff.

[the remark this answers is not in the document]
Comment (harrism) [reply cpp/include/cudf/fixed_point/temporary.hpp:20]: @bdice I went over all of your feedback and made sure it was addressed (except the things where I said "please open an issue".

File: cpp/src/io/json/legacy/reader_impl.cu [hunk 5]
```diff
@@ -14,14 +14,9 @@
  * limitations under the License.
  */
 
+#include "hash/concurrent_unordered_map.cuh"
```
[the remark this answers is not in the document]
Comment (harrism) [reply cpp/src/io/json/legacy/reader_impl.cu:17]: Yep.

File: cpp/src/io/parquet/reader_impl_helpers.cpp [hunk 6]
```diff
@@ -16,7 +16,7 @@
 
 #include "reader_impl_helpers.hpp"
 
-#include <io/utilities/row_selection.hpp>
+#include "io/utilities/row_selection.hpp"
```
[the remark this answers is not in the document]
Comment (harrism) [reply cpp/src/io/parquet/reader_impl_helpers.cpp:19]: Yep.

File: cpp/src/io/utilities/datasource.cpp [hunk 7]
```diff
@@ -21,14 +21,14 @@
...
-#include <io/utilities/config_utils.hpp>
 
 #include <kvikio/file_handle.hpp>
+
 #include <rmm/device_buffer.hpp>
 
 #include <arrow/io/memory.h>
-
 #include <fcntl.h>
+#include <io/utilities/config_utils.hpp>
```
[the remark this answers is not in the document]
Comment (harrism) [reply cpp/src/io/utilities/datasource.cpp]: Agree. Now done for all io, hash, join, and many other includes. Basically I searched for regexp `#include <(NAME/.*)>` and replaced with the same in quotes, where NAME is any folder name in `cudf/src`.

File: cpp/src/utilities/logger.cpp (hunk 3 above)
Comment (harrism) [reply cpp/src/utilities/logger.cpp]: Fixed, thx.

File: cpp/src/join/mixed_join_size_kernels_semi.cu (hunk 2 above)
Comment (harrism) [reply cpp/src/join/mixed_join_size_kernels_semi.cu]: Yes! Changed all `#include <SRCDIR/...>` to `#include "SRCDIR/..." where SRCDIR is any directory under `cudf/src`. From now on all includes of cudf headers that are not in `cudf/include` should be quoted. And the nice thing is when they are not, they end up down at the bottom of the includes, making it easier to spot.

We could consider quoting all includes of cuDF headers.

File: cpp/src/io/parquet/reader_impl.cpp (hunk 1 above)
Comment (harrism) [reply cpp/src/io/parquet/reader_impl.cpp:18]: Yep: https://github.com/rapidsai/cudf/pull/14993#discussion_r1483744225

**Comment by PointKernel** [comment]
> Other than headers from RAPIDS repos, I group all other headers that have a file extension in a single group, and all files that have no file extension in another group. Since the latter also matches includes some files from libcu++, I have an explicit category for <cuda/ includes to keep them separate from STL includes. A frequent effect of the single "." group is that cub and thrust headers get grouped without a blank line between them. I don't think this is a problem.


PR description to be updated. It seems all CCCL headers now belong to one header group.

**Comment by harrism** [comment]
> PR description to be updated. It seems all CCCL headers now belong to one header group.

Updated, thanks.

**Review by davidwendt — APPROVED** [review]
Looks good. Thanks for working on this.

**Review by bdice — APPROVED** [review]
Approving -- this PR is likely to have significant conflicts with others in flight, so I'd like to merge it ASAP. This is a significant improvement in the current state, and I'm sure we'll have more details to iron out later.

Thanks @harrism for addressing my feedback from the previous round. https://github.com/rapidsai/cudf/pull/15063#discussion_r1495103677

**Review by ttnghia — APPROVED** [review]

**Review by ttnghia — APPROVED** [review]

**Review by vyasr — COMMENTED** [review]
I didn't get a chance to review this earlier due to being out, but I agree with the approach and the results look good from a spot check of some of these files. Thanks for the ping, and thanks for doing this @harrism!

**Pull request merged**